### PR TITLE
display imagestream import errors

### PR DESCRIPTION
**Pull request by aweiteka**
To reproduce this error condition create new imagestream pointing to an insecure registry. The action to remedy this is to edit the imagestream selecting "insecure registry" (PR #4307).

screen shot 2016-04-29 at 4 20 18 pm

### CODE REVIEW

**Comment by aweiteka** [comment]
I forgot to add translatable attrs to these. I'll update when fixing any review feedback.

File: pkg/kubernetes/views/imagestream-body.html [hunk 1]
```diff
@@ -1,3 +1,14 @@
+<div ng-repeat="statustags in stream.status.tags">
+  <div ng-repeat="condition in statustags.conditions" ng-if='condition.status == "False"'
+      class="alert alert-danger alert-dismissable">
+    <button type="button" class="close" data-dismiss="alert" aria-hidden="true">
+      <span class="pficon pficon-close"></span>
+    </button>
+    <span class="pficon pficon-error-circle-o"></span>
+    <strong>{{ condition.reason }} on {{ condition.type }}:</strong> {{ condition.message }} {{ condition.lastTransitionTime }}
+      <a href="#" ng-click="modifyImageStream(stream)" class="alert-link">Edit imagestream</a>.
```
Comment (petervo) [inline pkg/kubernetes/views/imagestream-body.html]: Don't think we want to repeat this edit link for each one.

File: pkg/kubernetes/views/imagestream-body.html [hunk 2]
```diff
@@ -1,3 +1,14 @@
+<div ng-repeat="statustags in stream.status.tags">
+  <div ng-repeat="condition in statustags.conditions" ng-if='condition.status == "False"'
+      class="alert alert-danger alert-dismissable">
+    <button type="button" class="close" data-dismiss="alert" aria-hidden="true">
+      <span class="pficon pficon-close"></span>
+    </button>
+    <span class="pficon pficon-error-circle-o"></span>
+    <strong>{{ condition.reason }} on {{ condition.type }}:</strong> {{ condition.message }} {{ condition.lastTransitionTime }}
```
Comment (petervo) [inline pkg/kubernetes/views/imagestream-body.html]: This won't translate well, might need to be a filter.

Comment (aweiteka) [inline pkg/kubernetes/views/imagestream-body.html]: What do you suggest? How could I use a filter to address this?

File: pkg/kubernetes/views/imagestream-body.html (hunk 1 above)
Comment (aweiteka) [inline pkg/kubernetes/views/imagestream-body.html]: Yeah, ok. Maybe just display the $first element?

File: pkg/kubernetes/views/imagestream-body.html (hunk 2 above)
Comment (petervo) [inline pkg/kubernetes/views/imagestream-body.html]: It's probably ok for now, looking around there's a fair bit of this in the templates already. We'll need to fix at some point but for now it's probably ok.

**Comment by petervo** [comment]
Accidentally closed.

File: pkg/kubernetes/scripts/images.js [hunk 3]
```diff
@@ -356,6 +356,9 @@
              */
             select.register("taggedBy", function(tag) {
                 var i, len, results = { };
+                // catch condition when tag.items is null due to imagestream import error
+                if (!tag.items)
+                    tag.items = [];
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/images.js]: I think this should just ```return select(null);```. We shouldn't be modifying the tag after it's been loaded.

Comment (stefwalter) [inline pkg/kubernetes/scripts/images.js]: ```taggedFirst``` (above) also needs similar logic.

File: pkg/kubernetes/views/imagestream-body.html [hunk 4]
```diff
@@ -1,3 +1,16 @@
+<div ng-repeat="statustags in stream.status.tags">
+  <div ng-repeat="condition in statustags.conditions" ng-if='condition.status == "False"'
+      class="alert alert-danger alert-dismissable">
```
Comment (stefwalter) [inline pkg/kubernetes/views/imagestream-body.html]: Is this alert really dismissable? Does it disappear when the status changes?

File: pkg/kubernetes/views/imagestream-body.html (hunk 2 above)
Comment (stefwalter) [inline pkg/kubernetes/views/imagestream-body.html]: Do we really need the reason and the type? They don't seem to add anything to the user visible message.

File: pkg/kubernetes/views/imagestream-body.html [hunk 5]
```diff
@@ -1,3 +1,16 @@
...
+    <button type="button" class="close" data-dismiss="alert" aria-hidden="true">
+      <span class="pficon pficon-close"></span>
+    </button>
+    <span class="pficon pficon-error-circle-o"></span>
+    <strong>{{ condition.reason }} on {{ condition.type }}:</strong> {{ condition.message }} {{ condition.lastTransitionTime }}
+    <p ng-if='$first'>
+      <a translateable="yes" ng-click="modifyImageStream(stream)" class="alert-link">Edit imagestream</a>
```
Comment (stefwalter) [inline pkg/kubernetes/views/imagestream-body.html]: This is only a relevant course of action when it's an "ImportSuccess" condition right?

**Comment by stefwalter** [comment]
Do we also need a way to resync the imagestream? ie the equivalent of:

```
oc import-image name --loglevel=10
```

**Comment by stefwalter** [comment]
> Do we also need a way to resync the imagestream? ie the equivalent of:

Perhaps as a separate pull request?

File: pkg/kubernetes/views/imagestream-body.html (hunk 2 above)
Comment (aweiteka) [inline pkg/kubernetes/views/imagestream-body.html]: > Do we really need the reason and the type?

Probably not. Honestly I'm not clear on all the error conditions so I'm not sure how these might possibly be presented.

File: pkg/kubernetes/views/imagestream-body.html (hunk 4 above)
Comment (aweiteka) [inline pkg/kubernetes/views/imagestream-body.html]: > Is this alert really dismissable?

No, `alert-dismissable` class appears to be broken. @dperpeet you alerted me to this in IRC. I'm not sure how to debug. Any more details on this?

File: pkg/kubernetes/views/imagestream-body.html (hunk 5 above)
Comment (aweiteka) [inline pkg/kubernetes/views/imagestream-body.html]: Well, there aren't a lot of options for any error. Our choices are pretty much limited to delete and edit.

**Comment by aweiteka** [comment]
> Do we also need a way to resync the imagestream? ie the equivalent of:
> 
> oc import-image name --loglevel=10

Yes, I've been thinking that would be good to add. We need a bit of design work here. There are three use cases:

1. debugging (why is sync failing?)
1. initial discovery of remote tags (which ones are available for me to sync?)
1. manually updating (what images have changed in the upstream registry?)

**Comment by stefwalter** [comment]
> We need a bit of design work here

Worth a trello card?

**Comment by aweiteka** [comment]
> Worth a trello card?

https://trello.com/c/2OSvbawF/298-design-import-image-control

**Comment by aweiteka** [comment]
I think this is all set. Let me know if additional changes are needed.

@stefwalter

File: pkg/kubernetes/views/imagestream-body.html [hunk 6]
```diff
@@ -1,3 +1,16 @@
+<div ng-repeat="statustags in stream.status.tags">
+  <div ng-repeat="condition in statustags.conditions" ng-if='condition.status == "False"'
```
Comment (petervo) [inline pkg/kubernetes/views/imagestream-body.html]: I think we should target this to ImportSuccess conditions only, is that is the only implemented one right now. ei: ng-if="condition.type == 'ImportSuccess' && condition.status == "False"'. Makes the rest of this simpler, we we can remove the condition.reason and condition.type variables from the status string below, and remove the $first check from the edit image stream link

Comment (aweiteka) [inline pkg/kubernetes/views/imagestream-body.html]: I think you're right. I confirmed this based on the current types: https://github.com/openshift/origin/blob/master/pkg/image/api/v1/types.go#L139

File: pkg/kubernetes/views/imagestream-body.html (hunk 4 above)
Comment (aweiteka) [inline pkg/kubernetes/views/imagestream-body.html]: see #4401

**Pull request closed without merging**